text/3189-take-on-bool.md

```diff
@@ -0,0 +1,126 @@
+- Feature Name: `take-on-bool`
+- Start Date: 2021-10-27
+- RFC PR: [rust-lang/rfcs#3189](https://github.com/rust-lang/rfcs/pull/3189)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+This RFC proposes adding a method to `bool` called `take`, which will save the value of the bool to a variable, set the bool to false, and then return that variable.
+
+# Motivation
+
+In many applications, deferred execution of routines is preferable, especially around long standing state. Using flags, or booleans which indicate a condition, is a good way to raise those routines. Taking that flag, while also resetting it, makes this pattern more elegant, without complexifying the `std` inordinately.
+
+# Guide-level explanation
+
+All `.take` does is return the value of a boolean while also setting that value internally to `false`. It is exactly similar to `Option::take`, except that, of course, it only returns `true/false` instead of some inner value. (In this sense, this `.take` is effectively the same as `Option::take().is_some()`). In any example where flags are commonly read while also being reset, this method is useful.
+
+For example, imagine a common game structure:
+
+```rs
+/// A recursive data structure of positions, commonly used in games to allow a parent/child relationship between transforms.
+pub struct SceneNode {
+    /// a position relative to this Node's parent.
+    pub local_position: [f32; 2],
+    /// an authoritative position
+    pub world_position: [f32; 2],
+
+    pub dirty: bool,
+}
+
+impl SceneNode {
+    /// We want a new local position!
+    pub fn set_local_position(&mut self, new_pos: [f32; 2]) {
+        self.local_position = new_pos;
+        self.dirty = true;
+    }
+
+    /// we magically have the parent in this example.
+    pub fn calculuate_world_position(&mut self, parent: &SceneNode) {
+        // we can take the flag and also unset it in one method call
+        if self.dirty.take() {
+            self.world_position = [
+                self.local_position[0] + parent.local_position[0],
+                self.local_position[1] + parent.local_position[1],
+            ];
+        }
+
+        /*
+        // without this RFC, our code would be slightly uglier like this:
+        if self.dirty {
+            self.dirty = false;
+            self.world_position = [
+                self.local_position[0] + parent.local_position[0],
+                self.local_position[1] + parent.local_position[1],
+            ];
+        }
+        */
+    }
+}
+```
+
+# Reference-level explanation
+
+Implementation should be the following:
+
+```rs
+pub fn take(&mut self) -> bool {
+    // save the old value to a variable
+    let val = *self;
+    // and reset ourselves to false. If we are already false,
+    // then this doesn't matter.
+    *self = false;
```

Comment:
This should perhaps note that using `core::mem::take` does not actually optimize quite as well. It has an extra instruction compared:

```diff
         mov     al, byte ptr [rdi]
+        and     al, 1
         mov     byte ptr [rdi], 0
         ret
```
([Godbolt](https://rust.godbolt.org/z/5obcWaEKP))

Reply:
Yup -- I didn't want to get too in the weed here, but that's a great point. It mostly goes away once you call the method though

Reply:
That's not a reason to add this method, though, because it's a problem with every single two-variant enum, not just `bool`: <https://rust.godbolt.org/z/1Wjs1bf3Y>.

Its root cause is LLVM discarding information about the possible values of the discriminant.  <https://github.com/rust-lang/rust/issues/85133#issuecomment-904185574>  This "just" needs to get fixed, since it's actually hurting substantial amounts of `Option`- and `Result`-using code too.